### PR TITLE
Use a temporary file when updating the cache

### PR DESCRIPTION
**Pull request by eschnett**
This avoids conflicts if multiple Julia processes are updating the cache simultaneously.

### CODE REVIEW

File: base/loading.jl [hunk 1]
```diff
@@ -415,14 +415,15 @@ end
 evalfile(path::AbstractString, args::Vector) = evalfile(path, UTF8String[args...])
 
 function create_expr_cache(input::AbstractString, output::AbstractString)
-    isfile(output) && rm(output)
```
Comment (tkelman) [inline base/loading.jl]: no, this undoes https://github.com/JuliaLang/julia/pull/12723 - if precompilation fails then any existing ji file should also be removed

Comment (eschnett) [inline base/loading.jl]: Okay. Looking at the code again: Maybe all that is necessary is to change this line to
```
try rm(output) end
```
?

If `.ji` files are already created atomically, then the only remaining issue is that `rm` may fail if run in parallel.

Comment (tkelman) [inline base/loading.jl]: possibly, or maybe `rm` should be done in the `catch` below? let's see what the tests for #12720 say with the current version on CI

Comment (eschnett) [inline base/loading.jl]: Deleting first seems cleaner. This avoids the following scenario:
- Build 1 is started, fails, but takes a long time
- Build 2 starts later, succeeds, and produces a valid cache file
- Build 1 finally finishes, and deletes the newly created cache file

Deleting first ensures that the build that starts last either produces a cache file, or deletes it.

**Pull request closed without merging**